test_driver/transitions_perf_test.dart

```diff
@@ -0,0 +1,232 @@
+//
+// These names are reported by the test app, see _handleMessages()
+// in transitions_perf.dart.
+List<String> _allDemos = <String>[];
+
+// SerializableFinders for scrolling actions.
+final homeList = find.byValueKey('HomeListView');
+final backButton = find.byValueKey('Back');
+final galleryHeader = find.text('Gallery');
+final categoriesHeader = find.text('Categories');
+
+// Let overscroll animation settle on iOS after driver.scroll.
+void handleOverscrollAnimation() {
+  sleep(const Duration(seconds: 1));
+}
+
+/// Scroll to the top of the app, given the current demo. Works with both mobile
+/// and desktop layouts.
+Future scrollToTop(SerializableFinder demoItem, FlutterDriver driver) async {
+  print('scrolling to top');
+
+  // Scroll to the Categories header.
+  await driver.scroll(
+    demoItem,
+    0,
+    5000,
+    const Duration(milliseconds: 200),
+  );
+  handleOverscrollAnimation();
+
+  // Scroll to top.
+  await driver.scroll(
+    categoriesHeader,
+    0,
+    500,
+    const Duration(milliseconds: 200),
+  );
+  handleOverscrollAnimation();
+}
+
+/// Returns a [Future] that resolves to true if the widget specified by [finder]
+/// is present, false otherwise.
+Future<bool> isPresent(SerializableFinder finder, FlutterDriver driver,
+    {Duration timeout = const Duration(seconds: 5)}) async {
+  try {
+    await driver.waitFor(finder, timeout: timeout);
+    return true;
+  } catch (exception) {
+    return false;
+  }
+}
+
+/// Scrolls each each demo into view, launches it, then returns to the
+/// home screen, twice.
+Future<void> runDemos(List<String> demos, FlutterDriver driver) async {
+  String currentDemoCategory;
+  SerializableFinder demoList;
+  SerializableFinder demoItem;
+
+  for (final demo in demos) {
+    if (_skippedDemos.contains(demo)) continue;
+
+    print('> $demo');
+
+    final demoCategory = demo.substring(demo.indexOf('@') + 1);
+    if (demoCategory != currentDemoCategory) {
+      // We've switched categories.
+      currentDemoCategory = demoCategory;
+      demoList = find.byValueKey('${demoCategory}DemoList');
+
+      // We may want to return to the previous category later.
+      // Reset its scroll (matters for desktop layout).
+      if (demoItem != null) await scrollToTop(demoItem, driver);
+
+      // Scroll to the category list.
+      if (demoCategory != 'study') {
+        print('scrolling to $currentDemoCategory category');
+        await driver.scrollUntilVisible(
+          homeList,
+          demoList,
+          dyScroll: -1000,
+          timeout: const Duration(seconds: 10),
+        );
+      }
+    }
+
+    // Scroll to demo and open it twice.
+    demoItem = find.byValueKey(demo);
+
+    print('scrolling to demo');
+    await driver.scrollUntilVisible(
+      demoList,
+      demoItem,
+      dxScroll: -500,
+      dyScroll: -50,
+      alignment: 0.5,
+      timeout: const Duration(seconds: 10),
+    );
+
+    for (var i = 0; i < 2; i += 1) {
```

Comment:
Any reason that this is done exactly 2 times? Maybe add a comment explaining why